### PR TITLE
service reflection: include transitive closure for a file

### LINKED ISSUE #2949: service reflection: include transitive closure for a file
The [Java implementation](https://github.com/grpc/grpc-java/blob/master/services/src/main/java/io/grpc/protobuf/services/ProtoReflectionService.java#L298) of the reflection service includes the entire transitive closure for a request file, in toplogical order such that a file's dependencies always appear after the file (so the requested file is first, with all of its dependencies after it).

However, the Go implementation only ever includes a single file and never includes its dependencies. It would be nice to also include the transitive closure (at least if a given file has never been sent) so that clients have a better chance of trying to link files even if the import name differs from the descriptor name. This is basically the same problem as https://github.com/golang/protobuf/issues/567, except for downloading descriptors (vs. just querying the set of linked-in/registered descriptors).

Admittedly, having the set of transitive dependencies will still be error-prone if the names mismatch. The client could attempt to reconcile unlinkable descriptors by trying to match dependencies based only on the base name of the file and not its relative path. Or the set of descriptors could be provided in a well-specified order, that allows clients to compute the index for a given import and link it without relying on the name (which may not match).

For more context, see this: https://github.com/grpc/grpc/issues/19832

### PR DESCRIPTION
**Pull request by GarrettGutierrez1**
Fixes #2949.

### CODE REVIEW

File: reflection/serverreflection.go [hunk 1]
```diff
@@ -269,9 +269,41 @@ func (s *serverReflectionServer) allExtensionNumbersForType(st reflect.Type) ([]
...
+// wire format ([]byte). The fileDescriptors will include fd and all the
+// transitive dependencies of fd with names not in sentFileDescriptors.
+func fileDescWithDependencies(fd *dpb.FileDescriptorProto, sentFileDescriptors map[string]struct{}) ([][]byte, error) {
+	r := [][]byte{}
+	queue := make([]*dpb.FileDescriptorProto, 0)
+	queue = append(queue, fd)
+	for len(queue) > 0 {
+		currentfd := queue[0]
+		queue = queue[1:]
+		if _, exists := sentFileDescriptors[fd.GetName()]; len(r) == 0 || !exists {
```
Comment (dfawley) [inline reflection/serverreflection.go:282]: The "recommended" (by the Go team) pattern for implementing sets-as-maps in go is:

```go
map[string /* or whatever */]bool
if sent := sentFileDescriptors[fd.GetName()]; len(r) == 0 || !sent {
}
```

File: reflection/serverreflection.go [hunk 2]
```diff
@@ -269,9 +269,41 @@ func (s *serverReflectionServer) allExtensionNumbersForType(st reflect.Type) ([]
...
+			sentFileDescriptors[fd.GetName()] = struct{}{}
+			currentfdEncoded, err := proto.Marshal(currentfd)
+			if err != nil {
+				return nil, err
+			}
+			r = append(r, currentfdEncoded)
+
```
Comment (dfawley) [inline reflection/serverreflection.go:289]: Remove blank line please.

File: reflection/serverreflection.go [hunk 3]
```diff
@@ -269,9 +269,41 @@ func (s *serverReflectionServer) allExtensionNumbersForType(st reflect.Type) ([]
 	return out, nil
 }
 
+// fileDescWithDependencies returns a slice of serialized fileDescriptors in
+// wire format ([]byte). The fileDescriptors will include fd and all the
+// transitive dependencies of fd with names not in sentFileDescriptors.
+func fileDescWithDependencies(fd *dpb.FileDescriptorProto, sentFileDescriptors map[string]struct{}) ([][]byte, error) {
+	r := [][]byte{}
+	queue := make([]*dpb.FileDescriptorProto, 0)
+	queue = append(queue, fd)
```
Comment (dfawley) [inline reflection/serverreflection.go:278]:
```go
queue := []*dpb.FileDescriptorProto{fd}
```

File: reflection/serverreflection.go (hunk 1 above)
Comment (GarrettGutierrez1) [reply reflection/serverreflection.go:282]: Made this change.

File: reflection/serverreflection.go (hunk 2 above)
Comment (GarrettGutierrez1) [reply reflection/serverreflection.go:289]: Made this change.

File: reflection/serverreflection.go (hunk 3 above)
Comment (GarrettGutierrez1) [reply reflection/serverreflection.go:278]: Made this change.

**Comment by GarrettGutierrez1** [comment]
I got a simple server running with the C++ implementation and `grpcurl`/`grpc_cli` both have the same output with the go implementation. In none of the cases are the imported proto files reported in the output, so the output after this change is the same as the output prior to this change. That means a test on console output of `grpcurl`/`grpc_cli` can't verify this change. I could create a unit test that checks the messages sent back from the reflection service to ensure the imported proto file descriptors are sent.

**Comment by dfawley** [comment]
> I got a simple server running with the C++ implementation and grpcurl/grpc_cli both have the same output with the go implementation.

How simple is this service?

Does the Request/Response for your method include a proto message defined in another package?  Can you share your example and what you ran and what the output was?

**Comment by GarrettGutierrez1** [comment]
[Here](https://github.com/GarrettGutierrez1/grpc-simple) is the service I created to test. `proto/simple/simple.proto` imports `proto/simple/messages/messages.proto` (the latter implements the message types used by the former). Below are the outputs for grpc_cli and grpcurl.

```
$ grpc_cli ls localhost:50051 simple.Simple -l
filename: simple/simple.proto
package: simple;
service Simple {
  rpc SimpleUnary(simple.messages.SimpleRequest) returns (simple.messages.SimpleReply) {}
  rpc SimpleServerStream(simple.messages.SimpleRequest) returns (stream simple.messages.SimpleReply) {}
  rpc SimpleClientStream(stream simple.messages.SimpleRequest) returns (simple.messages.SimpleReply) {}
  rpc SimpleBidirectionalStream(stream simple.messages.SimpleRequest) returns (stream simple.messages.SimpleReply) {}
}

$ grpcurl -v -plaintext localhost:50051 list simple.Simple
simple.Simple.SimpleBidirectionalStream
simple.Simple.SimpleClientStream
simple.Simple.SimpleServerStream
simple.Simple.SimpleUnary
```

**Comment by dfawley** [comment]
You're just doing a `list` for both.  You'll need to actually try to send a request via `grpcurl`/`grpc_cli` to see what happens if it can't marshal the messages with the data the server is including in the reflection response.  My guess is `grpcurl` (maybe `grpc_cli`) will compensate for this by requesting the dependencies explicitly, but I don't know for sure.

**Comment by GarrettGutierrez1** [comment]
The output with grpcurl, on the go server, with the simple server provided is now as follows
```
$ grpcurl -plaintext -d '{"host": "myhost", "file_by_filename": "simple/simple.proto"}' localhost:50051 grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo
{
  "validHost": "myhost",
  "originalRequest": {
    "host": "myhost",
    "fileByFilename": "simple/simple.proto"
  },
  "fileDescriptorResponse": {
    "fileDescriptorProto": [
      "ChNzaW1wbGUvc2ltcGxlLnByb3RvEgZzaW1wbGUaHnNpbXBsZS9tZXNzYWdlcy9tZXNzYWdlcy5wcm90bzLoAgoGU2ltcGxlEk0KC1NpbXBsZVVuYXJ5Eh4uc2ltcGxlLm1lc3NhZ2VzLlNpbXBsZVJlcXVlc3QaHC5zaW1wbGUubWVzc2FnZXMuU2ltcGxlUmVwbHkiABJWChJTaW1wbGVTZXJ2ZXJTdHJlYW0SHi5zaW1wbGUubWVzc2FnZXMuU2ltcGxlUmVxdWVzdBocLnNpbXBsZS5tZXNzYWdlcy5TaW1wbGVSZXBseSIAMAESVgoSU2ltcGxlQ2xpZW50U3RyZWFtEh4uc2ltcGxlLm1lc3NhZ2VzLlNpbXBsZVJlcXVlc3QaHC5zaW1wbGUubWVzc2FnZXMuU2ltcGxlUmVwbHkiACgBEl8KGVNpbXBsZUJpZGlyZWN0aW9uYWxTdHJlYW0SHi5zaW1wbGUubWVzc2FnZXMuU2ltcGxlUmVxdWVzdBocLnNpbXBsZS5tZXNzYWdlcy5TaW1wbGVSZXBseSIAKAEwAUIsWipnaXRodWIuY29tL0dhcnJldHRHdXRpZXJyZXoxL3NpbXBsZS9zaW1wbGViBnByb3RvMw==",
      "Ch5zaW1wbGUvbWVzc2FnZXMvbWVzc2FnZXMucHJvdG8SD3NpbXBsZS5tZXNzYWdlcyIpCg1TaW1wbGVSZXF1ZXN0EhgKB21lc3NhZ2UYASABKAlSB21lc3NhZ2UiJwoLU2ltcGxlUmVwbHkSGAoHbWVzc2FnZRgBIAEoCVIHbWVzc2FnZUI1WjNnaXRodWIuY29tL0dhcnJldHRHdXRpZXJyZXoxL3NpbXBsZS9zaW1wbGUvbWVzc2FnZXNiBnByb3RvMw=="
    ]
  }
}
```
I see there are two file descriptors sent, however the 2nd one is different with the go server than with the C++ server? The first is identical in both.

**Comment by dfawley** [comment]
(Reassigning to reflect that we're waiting on more information about how this behaves when a normal RPC is called via `grpc_cli` / `grpcurl`), and what the differences observed mean (i.e. un-base64/gzip the FDProtos).

**Comment by GarrettGutierrez1** [comment]
File descriptor for messages.proto (which is imported) from the C++ server is as follows, followed by the go server. Looks like go just has a `json_name` parameter for the message types.

```
name:"simple/messages/messages.proto"
package:"simple.messages"
message_type:{name:"SimpleRequest"  field:{name:"message"  number:1  label:LABEL_OPTIONAL  type:TYPE_STRING}}
message_type:{name:"SimpleReply"  field:{name:"message"  number:1  label:LABEL_OPTIONAL  type:TYPE_STRING}}
options:{go_package:"github.com/GarrettGutierrez1/simple/simple/messages"}  syntax:"proto3"
```

```
name:"simple/messages/messages.proto"
package:"simple.messages"
message_type:{name:"SimpleRequest"  field:{name:"message"  number:1  label:LABEL_OPTIONAL  type:TYPE_STRING  json_name:"message"}}
message_type:{name:"SimpleReply"  field:{name:"message"  number:1  label:LABEL_OPTIONAL  type:TYPE_STRING  json_name:"message"}}
options:{go_package:"github.com/GarrettGutierrez1/simple/simple/messages"}  syntax:"proto3"
```

File: reflection/serverreflection_test.go [hunk 4]
```diff
@@ -227,6 +229,39 @@ func (x) TestReflectionEnd2end(t *testing.T) {
...
+		// io.EOF is not ok.
+		t.Fatalf("failed to recv response: %v", err)
+	}
+	switch r.MessageResponse.(type) {
+	case *rpb.ServerReflectionResponse_FileDescriptorResponse:
+		if !reflect.DeepEqual(r.GetFileDescriptorResponse().FileDescriptorProto[0], fdProto2Ext2Byte) {
+			t.Errorf("FileByFilename(%v)\nreceived: %q,\nwant: %q", filename, r.GetFileDescriptorResponse().FileDescriptorProto[0], fdProto2Ext2Byte)
+		}
+		if expectClosure {
+			if len(r.GetFileDescriptorResponse().FileDescriptorProto) < 2 {
```
Comment (dfawley) [inline reflection/serverreflection_test.go:252]: `!= 2` would be even more precise.

**Review by dfawley — APPROVED** [review]

**Pull request merged**